src/Compilers/Core/Portable/SourceGeneration/Nodes/NodeStateTable.cs

```diff
@@ -174,7 +174,7 @@ public sealed class Builder
 
             internal Builder(NodeStateTable<T> previous, string? name, bool stepTrackingEnabled)
             {
-                _states = ArrayBuilder<TableEntry>.GetInstance();
+                _states = ArrayBuilder<TableEntry>.GetInstance(previous._states.Length);
```

Comment:
note: this pooled array is still highly problematic.  because it routinely is >128 elements, we never actually pool it.  so we end up churning GB of arrays in incremental scenarios.

Reply:
Would creating a dedicated pool for generators that deals in large arrays only be a potential help here?